File: sdks/php/src/Api/ApiKeys.php
<?php

declare(strict_types=1);

namespace Apollo\Deploy\Signal\Sdk\Api;

use Apollo\Deploy\Signal\Sdk\SDKError;
use Apollo\Deploy\Signal\Sdk\SSEEvent;
use Apollo\Deploy\Signal\Sdk\Transport;
use Apollo\Deploy\Signal\Sdk\ApiKey;
use Apollo\Deploy\Signal\Sdk\ApiKeyUsageResponse;
use Apollo\Deploy\Signal\Sdk\ListApiKeysResponse;
use Psr\Http\Message\ResponseInterface;

/** ApiKeysAPI domain operations. */
class ApiKeys
{
    public function __construct(private Transport $transport)
    {
    }

    /**
     * @param string $projectId
     * @return ListApiKeysResponse
     * @throws SDKError
     */
    public function listApiKeys(
        string $projectId,
    ): ListApiKeysResponse
    {
        $path = '/v1/projects/${projectId}/api-keys';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ListApiKeysResponse::class, 'fromArray')) {
            return ListApiKeysResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $keyId
     * @return ApiKey
     * @throws SDKError
     */
    public function getApiKey(
        string $projectId,
        string $keyId,
    ): ApiKey
    {
        $path = '/v1/projects/${projectId}/api-keys/${keyId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${keyId}', rawurlencode(Transport::wireValue($keyId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ApiKey::class, 'fromArray')) {
            return ApiKey::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $keyId
     * @return ApiKeyUsageResponse
     * @throws SDKError
     */
    public function getApiKeyUsage(
        string $projectId,
        string $keyId,
    ): ApiKeyUsageResponse
    {
        $path = '/v1/projects/${projectId}/api-keys/${keyId}/usage';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${keyId}', rawurlencode(Transport::wireValue($keyId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ApiKeyUsageResponse::class, 'fromArray')) {
            return ApiKeyUsageResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $keyId
     * @return void
     * @throws SDKError
     */
    public function exportApiKeyUsage(
        string $projectId,
        string $keyId,
    ): void
    {
        $path = '/v1/projects/${projectId}/api-keys/${keyId}/usage/export';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${keyId}', rawurlencode(Transport::wireValue($keyId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    private function decodeResponse(ResponseInterface $response): mixed
    {
        $contents = $this->transport->readResponseBody($response);
        if (trim($contents) === '') {
            $requestId = $response->getHeaderLine('x-request-id')
                ?: $response->getHeaderLine('request-id')
                ?: null;
            throw new SDKError(
                message: 'Expected a response body for a non-void operation',
                status: $response->getStatusCode(),
                code: 'RESPONSE_BODY_MISSING',
                requestId: $requestId,
            );
        }

        try {
            return json_decode($contents, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return $contents;
        }
    }
}

File: sdks/rust/src/api/api_keys.rs
use crate::error::SdkError;
use crate::transport::Transport;
use reqwest::Method;
use serde::Serialize;
use crate::types::{ ApiKey, ApiKeyUsageResponse, ListApiKeysResponse };

/// ApiKeysAPI API operations.
pub struct ApiKeysAPI {
    transport: Transport,
}

impl ApiKeysAPI {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub async fn list_api_keys(
        &self,
        project_id: String,
    ) -> Result<ListApiKeysResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/api-keys".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ListApiKeysResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_api_key(
        &self,
        project_id: String,
        key_id: String,
    ) -> Result<ApiKey, SdkError> {
        let mut path = "/v1/projects/${projectId}/api-keys/${keyId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${keyId}", &encode_path(&key_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ApiKey>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_api_key_usage(
        &self,
        project_id: String,
        key_id: String,
    ) -> Result<ApiKeyUsageResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/api-keys/${keyId}/usage".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${keyId}", &encode_path(&key_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ApiKeyUsageResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn export_api_key_usage(
        &self,
        project_id: String,
        key_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/api-keys/${keyId}/usage/export".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${keyId}", &encode_path(&key_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

}

fn encode_path(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

fn encode_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    match serde_json::to_value(value)
        .map_err(|error| SdkError::ValidationError(error.to_string()))?
    {
        serde_json::Value::String(value) => Ok(value),
        value => Ok(value.to_string()),
    }
}

fn encode_cookie_name(name: &str) -> String {
    urlencoding::encode(name).into_owned()
}

fn encode_cookie_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    let value = encode_value(value)?;
    Ok(urlencoding::encode(&value).into_owned())
}
